src/Workspaces/Core/Portable/Workspace/Solution/SourceGeneratedDocumentState.cs

```diff
@@ -192,23 +227,34 @@ public SourceGeneratedDocumentState WithGenerationDateTime(DateTime generationDa
     public SourceGeneratedDocumentState WithSyntaxRoot(SyntaxNode newRoot)
     {
         // See if we can reuse this instance directly
-        if (this.TryGetSyntaxTree(out var tree) && tree == newRoot.SyntaxTree)
+        if (this.TryGetSyntaxTree(out var tree) &&
+            tree.TryGetRoot(out var root) &&
+            root == newRoot)
+        {
             return this;
+        }
+
+        var sourceText = newRoot.GetText();
```

Comment:
It still feels to me like we could be avoiding this call if we used the CreateTreeWithLazyText helper in DocumentState.cs, but maybe it's fine. I don't think this is actually realizing the text, right? It's just returning a text that walks the tree as needed?

Reply:
We could call that, yes, but at this point the `SourceGeneratedDocumentState` still needs a fully realised `SourceText` at construction time. Making it fully lazy is beyond this PR IMO. I haven't looked at how the `SourceText` property is used, but there is probably a reasonably easy path to, if not fully lazy, at least lazy-as-long-as-its-synchronous.